osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs

```diff
@@ -104,55 +104,55 @@ public OsuLegacySkinTransformer(ISkin skin)
                             if (followCircleContent != null)
                                 return new LegacyFollowCircle(followCircleContent);
 
-                            return null;
+                            break;
 
                         case OsuSkinComponents.SliderBall:
                             if (GetTexture("sliderb") != null || GetTexture("sliderb0") != null)
                                 return new LegacySliderBall(this);
 
-                            return null;
+                            break;
```

Comment:
So some discussing in twitch chat took place (most of which was me floundering about talking how I don't get any of this) but one of the main *productive* discussion points that fell out of that discussion was that this sort of change baffles me.

Unless I'm grossly misunderstanding the structure (and that *is* possible): This is a legacy skin transformer for osu! specifically. The skin doesn't have the `sliderb` texture ---- so why is this not blocking any further fallbacks, knowing that they *cannot* be serviced sanely, and instead falling back to the base, which might do *anything* as far as I'm concerned, including the wrong thing of returning some bogus component rather than stopping the lookup *for that given skin* and instead letting the next skin handle it?

Reply:
The reason I made this PR was to address the concerns that I perceived at least one of us had in IRL discussions, by standardising all calls to `base.GetDrawableComponent` rather than `null`. I'd ask that @frenzibyte and @smoogipoo check this PR and give their take on direction.

I tend to prefer having the `null` returns for cases where "the base thing is never going to be able to service this anyway", but I seem to recollect from our discussions that this was not a unanimous agreement. My brain can work with either, so needs more opinions.

Reply:
There are three transformations:
1. Returning a totally new component with no correlation to the original (variations of `return new ...`).
2. Transforming the original component (`return base.GetDrawableComponent(...).With(...)`).
3. Blocking a component (`return null`).

Most cases fall under (1) or (2).

Blocking a component from being returned is a very special case (to me) because it requires having white-box knowledge of the underlying skin and what components it returns. While yes, we (the limited set of developers involved in skinning thus far) have this knowledge, it begs the question of whether we need the knowledge in the first place.

If we assume that, in general, `DefaultLegacySkin`, `LegacySkin`, `ArgonSkin`, `TrianglesSkin` all do not have knowledge of skin-specific components because they're part of the `osu.Game` namespace, then do we need to bother with the difference between calling base (still required in some cases) and returning null, if there even is any?

So I think for the most part I proposed this to dumb down the entire thing and treat `return base.GetDrawableComponent()` equivalent to a `default` case, rather than that being the point of `null`. As it is, because the `base` call exists (and has to exist in some cases), it leads the programmer to believe there's something special going on with the `null` returns.  
You can claim that there _is_ something special by "knowing that they cannot be serviced sanely", but I think that is a weak argument that lends credibility to my claim of white-box knowledge.

All of that being said, I'm sure @peppy will write a few words/close this PR in the next comment, because my conviction for this is only on the order of 51%/49%.

Reply:
Yeah for now I'll just close this one. Will approach the next step forward from a different direction.

Reply:
> If we assume that, in general, `DefaultLegacySkin`, `LegacySkin`, `ArgonSkin`, `TrianglesSkin` all do not have knowledge of skin-specific components because they're part of the `osu.Game` namespace, then do we need to bother with the difference between calling base (still required in some cases) and returning null, if there even is any?

The case I pointed out above is in a *skin transformer* living in `osu.Game.Rulesets.Osu`. I can't imagine designing sanely in a ruleset project from the baseline of "not having knowledge of skin-specific components".

Reply:
Thinking about this again, since there mustn't be a difference between using `break` or `return null` here, I would personally favour using `return null` as much as possible than `break`, because it makes me mentally imagine the `GetDrawableComponent` lookup sequence to be as short as possible. I nodded to using `break` everywhere at first, but reading the full diff now, I can understand spaceman's concern here.

I would want `base.GetDrawableComponent` to only be called when no type is matched against the given lookup, otherwise aggressively return null. 

The way I imagine `GetDrawableComponent` to work in general is that each implementation of the method checks the lookup against a specific type, if it matches that, it either returns the requested component or null, otherwise fall back to base implementation.

This impacts usages such as [this](https://github.com/ppy/osu/blob/36b4013fa64857c23c70ab8f591bc2cc6b18c44f/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs#L32-L37) and [that](https://github.com/ppy/osu/blob/48cfd77ee8d0ef359db019855ce6653103b23cef/osu.Game/Skinning/ArgonSkin.cs#L94-L95). I would like it that we handle this completely differently, because it's honestly just hacky in my eyes, and complicates the purpose of transformers, which I'm pretty sure was meant to just add ruleset-specific components support, not transform existing components. I also propose removing `SpriteComponentLookup` as a result of this, I do not like that there's "default handling" for this lookup type or any lookup type at **Skin** itself (user customisation is an exception, because `Skin` itself is the one introducing the concept of user customisation anyway). I believe that this sort of handling should happen at `SkinnableSprite` itself, `Skin` should not be concerned about it.

I've tried to put my vision on this in a diff, but one thing that will probably get me yelled at is my usage of `case ... when ...`. The sole purpose why I'm doing that is to achieve zero usages of `break`. Open to discussion.

Here's the diff anyway:
<details>
<summary>diff</summary>

```diff
diff --git a/osu.Game.Rulesets.Catch/Skinning/Argon/CatchArgonSkinTransformer.cs b/osu.Game.Rulesets.Catch/Skinning/Argon/CatchArgonSkinTransformer.cs
index 520c2de248..be8a3234b9 100644
--- a/osu.Game.Rulesets.Catch/Skinning/Argon/CatchArgonSkinTransformer.cs
+++ b/osu.Game.Rulesets.Catch/Skinning/Argon/CatchArgonSkinTransformer.cs
@@ -18,7 +18,6 @@ public CatchArgonSkinTransformer(ISkin skin)
             switch (lookup)
             {
                 case CatchSkinComponentLookup catchComponent:
-                    // TODO: Once everything is finalised, consider throwing UnsupportedSkinComponentException on missing entries.
                     switch (catchComponent.Component)
                     {
                         case CatchSkinComponents.HitExplosion:
@@ -35,9 +34,13 @@ public CatchArgonSkinTransformer(ISkin skin)
 
                         case CatchSkinComponents.Droplet:
                             return new ArgonDropletPiece();
-                    }
 
-                    break;
+                        case CatchSkinComponents.CatchComboCounter:
+                            return null; // not implemented.
+
+                        default:
+                            throw new UnsupportedSkinComponentException(lookup);
+                    }
             }
 
             return base.GetDrawableComponent(lookup);
diff --git a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
index 69efb7fbca..6e6f381422 100644
--- a/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Catch/Skinning/Legacy/CatchLegacySkinTransformer.cs
@@ -30,11 +30,7 @@ public CatchLegacySkinTransformer(ISkin skin)
         {
             switch (lookup)
             {
-                case GlobalSkinnableContainerLookup containerLookup:
-                    // Only handle per ruleset defaults here.
-                    if (containerLookup.Ruleset == null)
-                        return base.GetDrawableComponent(lookup);
-
+                case GlobalSkinnableContainerLookup containerLookup when containerLookup.Ruleset != null:
                     // we don't have enough assets to display these components (this is especially the case on a "beatmap" skin).
                     if (!IsProvidingLegacyResources)
                         return null;
diff --git a/osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs b/osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs
index afccb2e568..0978f434c4 100644
--- a/osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs
+++ b/osu.Game.Rulesets.Mania/Skinning/Argon/ManiaArgonSkinTransformer.cs
@@ -28,11 +28,7 @@ public ManiaArgonSkinTransformer(ISkin skin, IBeatmap beatmap)
         {
             switch (lookup)
             {
-                case GlobalSkinnableContainerLookup containerLookup:
-                    // Only handle per ruleset defaults here.
-                    if (containerLookup.Ruleset == null)
-                        return base.GetDrawableComponent(lookup);
-
+                case GlobalSkinnableContainerLookup containerLookup when containerLookup.Ruleset != null:
                     switch (containerLookup.Lookup)
                     {
                         case GlobalSkinnableContainers.MainHUDComponents:
@@ -63,7 +59,6 @@ public ManiaArgonSkinTransformer(ISkin skin, IBeatmap beatmap)
                     return new ArgonJudgementPiece(resultComponent.Component);
 
                 case ManiaSkinComponentLookup maniaComponent:
-                    // TODO: Once everything is finalised, consider throwing UnsupportedSkinComponentException on missing entries.
                     switch (maniaComponent.Component)
                     {
                         case ManiaSkinComponents.StageBackground:
@@ -92,9 +87,14 @@ public ManiaArgonSkinTransformer(ISkin skin, IBeatmap beatmap)
 
                         case ManiaSkinComponents.HitExplosion:
                             return new ArgonHitExplosion();
-                    }
 
-                    break;
+                        case ManiaSkinComponents.StageForeground:
+                        case ManiaSkinComponents.BarLine:
+                            return null; // not implemented.
+
+                        default:
+                            throw new UnsupportedSkinComponentException(lookup);
+                    }
             }
 
             return base.GetDrawableComponent(lookup);
diff --git a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs b/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
index cb42b2b62a..2ce4f623e0 100644
--- a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs
@@ -80,11 +80,7 @@ public override Drawable GetDrawableComponent(ISkinComponentLookup lookup)
         {
             switch (lookup)
             {
-                case GlobalSkinnableContainerLookup containerLookup:
-                    // Modifications for global components.
-                    if (containerLookup.Ruleset == null)
-                        return base.GetDrawableComponent(lookup);
-
+                case GlobalSkinnableContainerLookup containerLookup when containerLookup.Ruleset != null:
                     // we don't have enough assets to display these components (this is especially the case on a "beatmap" skin).
                     if (!IsProvidingLegacyResources)
                         return null;
diff --git a/osu.Game.Rulesets.Osu/Skinning/Argon/OsuArgonSkinTransformer.cs b/osu.Game.Rulesets.Osu/Skinning/Argon/OsuArgonSkinTransformer.cs
index 9f6f65c206..b9430ccb4b 100644
--- a/osu.Game.Rulesets.Osu/Skinning/Argon/OsuArgonSkinTransformer.cs
+++ b/osu.Game.Rulesets.Osu/Skinning/Argon/OsuArgonSkinTransformer.cs
@@ -36,7 +36,6 @@ public OsuArgonSkinTransformer(ISkin skin)
                     }
 
                 case OsuSkinComponentLookup osuComponent:
-                    // TODO: Once everything is finalised, consider throwing UnsupportedSkinComponentException on missing entries.
                     switch (osuComponent.Component)
                     {
                         case OsuSkinComponents.HitCircle:
@@ -71,9 +70,19 @@ public OsuArgonSkinTransformer(ISkin skin)
 
                         case OsuSkinComponents.CursorTrail:
                             return new ArgonCursorTrail();
-                    }
 
-                    break;
+                        case OsuSkinComponents.CursorParticles:
+                        case OsuSkinComponents.CursorRipple:
+                        case OsuSkinComponents.HitCircleText:
+                        case OsuSkinComponents.SliderTailHitCircle:
+                        case OsuSkinComponents.CursorSmoke:
+                        case OsuSkinComponents.ApproachCircle:
+                            // not implemented.
+                            return null;
+
+                        default:
+                            throw new UnsupportedSkinComponentException(lookup);
+                    }
             }
 
             return base.GetDrawableComponent(lookup);
diff --git a/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs b/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
index 636a9ecb21..70ee85ba62 100644
--- a/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinTransformer.cs
@@ -44,11 +44,7 @@ public OsuLegacySkinTransformer(ISkin skin)
         {
             switch (lookup)
             {
-                case GlobalSkinnableContainerLookup containerLookup:
-                    // Only handle per ruleset defaults here.
-                    if (containerLookup.Ruleset == null)
-                        return base.GetDrawableComponent(lookup);
-
+                case GlobalSkinnableContainerLookup containerLookup when containerLookup.Ruleset != null:
                     // we don't have enough assets to display these components (this is especially the case on a "beatmap" skin).
                     if (!IsProvidingLegacyResources)
                         return null;
diff --git a/osu.Game.Rulesets.Taiko/Skinning/Argon/TaikoArgonSkinTransformer.cs b/osu.Game.Rulesets.Taiko/Skinning/Argon/TaikoArgonSkinTransformer.cs
index bfc9e8648d..43ebd710ed 100644
--- a/osu.Game.Rulesets.Taiko/Skinning/Argon/TaikoArgonSkinTransformer.cs
+++ b/osu.Game.Rulesets.Taiko/Skinning/Argon/TaikoArgonSkinTransformer.cs
@@ -26,7 +26,6 @@ public TaikoArgonSkinTransformer(ISkin skin)
                     return new ArgonJudgementPiece(resultComponent.Component);
 
                 case TaikoSkinComponentLookup taikoComponent:
-                    // TODO: Once everything is finalised, consider throwing UnsupportedSkinComponentException on missing entries.
                     switch (taikoComponent.Component)
                     {
                         case TaikoSkinComponents.CentreHit:
@@ -70,9 +69,15 @@ public TaikoArgonSkinTransformer(ISkin skin)
 
                         case TaikoSkinComponents.Swell:
                             return new ArgonSwellCirclePiece();
-                    }
 
-                    break;
+                        case TaikoSkinComponents.Scroller:
+                        case TaikoSkinComponents.Mascot:
+                        case TaikoSkinComponents.KiaiGlow:
+                            return null; // not implemented.
+
+                        default:
+                            throw new UnsupportedSkinComponentException(lookup);
+                    }
             }
 
             return base.GetDrawableComponent(lookup);
diff --git a/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs b/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs
index 5bdb824f1c..a41144dec0 100644
--- a/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs
+++ b/osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs
@@ -29,118 +29,122 @@ public TaikoLegacySkinTransformer(ISkin skin)
 
         public override Drawable? GetDrawableComponent(ISkinComponentLookup lookup)
         {
-            if (lookup is SkinComponentLookup<HitResult>)
+            switch (lookup)
             {
-                // if a taiko skin is providing explosion sprites, hide the judgements completely
-                if (hasExplosion.Value)
-                    return Drawable.Empty().With(d => d.Expire());
-            }
+                case SkinComponentLookup<HitResult> hitResultLookup:
+                    if (hasExplosion.Value)
+                    {
+                        // if a taiko skin is providing explosion sprites, hide the judgements completely
+                        return Drawable.Empty().With(d => d.Expire());
+                    }
 
-            if (lookup is TaikoSkinComponentLookup taikoComponent)
-            {
-                switch (taikoComponent.Component)
-                {
-                    case TaikoSkinComponents.DrumRollBody:
-                        if (GetTexture("taiko-roll-middle") != null)
-                            return new LegacyDrumRoll();
+                    // otherwise return base judgement implementations.
+                    return base.GetDrawableComponent(hitResultLookup);
 
-                        return null;
+                case TaikoSkinComponentLookup taikoComponent:
+                    switch (taikoComponent.Component)
+                    {
+                        case TaikoSkinComponents.DrumRollBody:
+                            if (GetTexture("taiko-roll-middle") != null)
+                                return new LegacyDrumRoll();
 
-                    case TaikoSkinComponents.InputDrum:
-                        if (hasBarLeft)
-                            return new LegacyInputDrum();
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.InputDrum:
+                            if (hasBarLeft)
+                                return new LegacyInputDrum();
 
-                    case TaikoSkinComponents.DrumSamplePlayer:
-                        return null;
+                            return null;
 
-                    case TaikoSkinComponents.CentreHit:
-                    case TaikoSkinComponents.RimHit:
-                        if (hasHitCircle)
-                            return new LegacyHit(taikoComponent.Component);
+                        case TaikoSkinComponents.DrumSamplePlayer:
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.CentreHit:
+                        case TaikoSkinComponents.RimHit:
+                            if (hasHitCircle)
+                                return new LegacyHit(taikoComponent.Component);
 
-                    case TaikoSkinComponents.DrumRollTick:
-                        return this.GetAnimation("sliderscorepoint", false, false);
+                            return null;
 
-                    case TaikoSkinComponents.Swell:
-                        // todo: support taiko legacy swell (https://github.com/ppy/osu/issues/13601).
-                        return null;
+                        case TaikoSkinComponents.DrumRollTick:
+                            return this.GetAnimation("sliderscorepoint", false, false);
 
-                    case TaikoSkinComponents.HitTarget:
-                        if (GetTexture("taikobigcircle") != null)
-                            return new TaikoLegacyHitTarget();
+                        case TaikoSkinComponents.Swell:
+                            // todo: support taiko legacy swell (https://github.com/ppy/osu/issues/13601).
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.HitTarget:
+                            if (GetTexture("taikobigcircle") != null)
+                                return new TaikoLegacyHitTarget();
 
-                    case TaikoSkinComponents.PlayfieldBackgroundRight:
-                        if (GetTexture("taiko-bar-right") != null)
-                            return new TaikoLegacyPlayfieldBackgroundRight();
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.PlayfieldBackgroundRight:
+                            if (GetTexture("taiko-bar-right") != null)
+                                return new TaikoLegacyPlayfieldBackgroundRight();
 
-                    case TaikoSkinComponents.PlayfieldBackgroundLeft:
-                        // This is displayed inside LegacyInputDrum. It is required to be there for layout purposes (can be seen on legacy skins).
-                        if (GetTexture("taiko-bar-right") != null)
-                            return Drawable.Empty();
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.PlayfieldBackgroundLeft:
+                            // This is displayed inside LegacyInputDrum. It is required to be there for layout purposes (can be seen on legacy skins).
+                            if (GetTexture("taiko-bar-right") != null)
+                                return Drawable.Empty();
 
-                    case TaikoSkinComponents.BarLine:
-                        if (GetTexture("taiko-barline") != null)
-                            return new LegacyBarLine();
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.BarLine:
+                            if (GetTexture("taiko-barline") != null)
+                                return new LegacyBarLine();
 
-                    case TaikoSkinComponents.TaikoExplosionMiss:
-                        var missSprite = this.GetAnimation(getHitName(taikoComponent.Component), true, false);
-                        if (missSprite != null)
-                            return new LegacyHitExplosion(missSprite);
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.TaikoExplosionMiss:
+                            var missSprite = this.GetAnimation(getHitName(taikoComponent.Component), true, false);
+                            if (missSprite != null)
+                                return new LegacyHitExplosion(missSprite);
 
-                    case TaikoSkinComponents.TaikoExplosionOk:
-                    case TaikoSkinComponents.TaikoExplosionGreat:
-                        string hitName = getHitName(taikoComponent.Component);
-                        var hitSprite = this.GetAnimation(hitName, true, false);
+                            return null;
 
-                        if (hitSprite != null)
-                        {
-                            var strongHitSprite = this.GetAnimation($"{hitName}k", true, false);
+                        case TaikoSkinComponents.TaikoExplosionOk:
+                        case TaikoSkinComponents.TaikoExplosionGreat:
+                            string hitName = getHitName(taikoComponent.Component);
+                            var hitSprite = this.GetAnimation(hitName, true, false);
 
-                            return new LegacyHitExplosion(hitSprite, strongHitSprite);
-                        }
+                            if (hitSprite != null)
+                            {
+                                var strongHitSprite = this.GetAnimation($"{hitName}k", true, false);
 
-                        return null;
+                                return new LegacyHitExplosion(hitSprite, strongHitSprite);
+                            }
 
-                    case TaikoSkinComponents.TaikoExplosionKiai:
-                        // suppress the default kiai explosion if the skin brings its own sprites.
-                        // the drawable needs to expire as soon as possible to avoid accumulating empty drawables on the playfield.
-                        if (hasExplosion.Value)
-                            return Drawable.Empty().With(d => d.Expire());
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.TaikoExplosionKiai:
+                            // suppress the default kiai explosion if the skin brings its own sprites.
+                            // the drawable needs to expire as soon as possible to avoid accumulating empty drawables on the playfield.
+                            if (hasExplosion.Value)
+                                return Drawable.Empty().With(d => d.Expire());
 
-                    case TaikoSkinComponents.Scroller:
-                        if (GetTexture("taiko-slider") != null)
-                            return new LegacyTaikoScroller();
+                            return null;
 
-                        return null;
+                        case TaikoSkinComponents.Scroller:
+                            if (GetTexture("taiko-slider") != null)
+                                return new LegacyTaikoScroller();
 
-                    case TaikoSkinComponents.Mascot:
-                        return new DrawableTaikoMascot();
+                            return null;
 
-                    case TaikoSkinComponents.KiaiGlow:
-                        if (GetTexture("taiko-glow") != null)
-                            return new LegacyKiaiGlow();
+                        case TaikoSkinComponents.Mascot:
+                            return new DrawableTaikoMascot();
 
-                        return null;
+                        case TaikoSkinComponents.KiaiGlow:
+                            if (GetTexture("taiko-glow") != null)
+                                return new LegacyKiaiGlow();
 
-                    default:
-                        throw new UnsupportedSkinComponentException(lookup);
-                }
+                            return null;
+
+                        default:
+                            throw new UnsupportedSkinComponentException(lookup);
+                    }
             }
 
             return base.GetDrawableComponent(lookup);
```
</details>

Reply:
Worth noting that none of the above is meant to block moving in any direction, I'm absolutely fine with null or base call, I just stated my personal opinion on the matter.

Reply:
> which I'm pretty sure was meant to just add ruleset-specific components support, not transform existing components

To me, this is false and demonstrated in your diff via `TaikoLegacySkinTransformer` and its handling of `SkinComponentLookup<HitResult>`. This is also likely the same for `OsuTrianglesSkinTransformer`. If there is even _one_ case where this is not a "something or nothing" scenario, then it should be treated as a proper transformer - the fact that most cases resolve into `return null` is only a degenerate case that we've white-boxed ourselves into with the prior knowledge that no other implementation can handle the ruleset specific components.

But, I'll go with your thoughts...

> I believe that this sort of handling should happen at SkinnableSprite itself, Skin should not be concerned about it.

Okay, so is this how you see it being handled then?

```csharp
public override Drawable? GetSpriteComponent(string lookup, Vector2? maxSize)
{
    switch (lookup)
    {
        case "lighting":
            return null;

        default:
            return this.GetAnimation(lookup, false, false, maxSize: maxSize);
    }
}
```

What about hit results - do we have another `virtual` method for that? Does every `Component` lookup at some point resolve to a `virtual` method of its own? That is...

```csharp
public class MySkinTransformer : SkinTransformer,
                                 IComponentProvider<SkinComponentLookup<HitResult>>,
                                 IComponentProvider<SkinnableSprite.SpriteComponentLookup>,
                                 IComponentProvider<OsuSkinComponentLookup>
{
    public MySkinTransformer(ISkin skin)
        : base(skin)
    {
    }

    public Drawable? GetComponent(SkinComponentLookup<HitResult> lookup)
    {
        throw new System.NotImplementedException();
    }

    public Drawable? GetComponent(SkinnableSprite.SpriteComponentLookup lookup)
    {
        throw new System.NotImplementedException();
    }

    public Drawable? GetComponent(OsuSkinComponentLookup lookup)
    {
        throw new System.NotImplementedException();
    }
}

public interface IComponentProvider<in TComponent>
    where TComponent : ISkinComponentLookup
{
    Drawable? GetComponent(TComponent lookup);
}
```

Taking that idea further, we can extend it with
```csharp
public class SampleComponentLookup : ISkinComponentLookup
{
    public ISampleInfo Sample { get; }
}
```
To handle overrides of [`GetSample`](https://github.com/ppy/osu/blob/ccf1acce56798497edfaf92d3ece933469edcf0a/osu.Game.Rulesets.Mania/Skinning/Legacy/ManiaLegacySkinTransformer.cs#L177). Same could be done for textures and config, and maybe everything could be unified under "component lookup"s, with subsequent removal of `SkinTransformer`, `ISkinTransformer`, and passing `ISkin` via that `GetComponent()` method.

I'll give this idea a shot just to see how it would look in a practical scenario.

Reply:
So I did a thing... (probably gone crazy is what)

https://github.com/ppy/osu/compare/master...smoogipoo:osu:skin-component-provider-2?expand=1

I want to preface by saying I'm not trying to push this or anything, it's just a deep dive to play around with things. Some interesting tidbits:
- [`OsuLegacySkinComponentProvider`](https://github.com/smoogipoo/osu/blob/65e157d84171f1c42ca33be55d7fd63ccac63ace/osu.Game.Rulesets.Osu/Skinning/Legacy/OsuLegacySkinComponentProvider.cs), previously `OsuLegacySkinTransformer`.
- [`LegacySkin_Providers.cs`](https://github.com/smoogipoo/osu/blob/65e157d84171f1c42ca33be55d7fd63ccac63ace/osu.Game/Skinning/LegacySkin_Providers.cs), previously the implementation of `LegacySkin`. Note also `LegacySkinTransformer.GetSample()` is merged into this implementation.
- Some hackery in [`SkinnableDrawable`](https://github.com/smoogipoo/osu/blob/65e157d84171f1c42ca33be55d7fd63ccac63ace/osu.Game/Skinning/SkinnableDrawable.cs#L79-L83), would eventually require `SkinnableDrawable<T:ISkinComponentLookup>`.
- It's probably way too complex than it's worth, even if it's being more explicit about things.

Will stop here for now and let my mind stew on what I actually want to achieve, if anything :)